packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,641 @@
+  DEFAULT_BUSINESS_DATE_MILESTONING_PARAMETER_NAME,
+} from '@finos/legend-graph';
+import {
+  fillDerivedPropertyArguments,
+  type QueryBuilderPropertyExpressionState,
+  type QueryBuilderDerivedPropertyExpressionState,
+} from './QueryBuilderPropertyEditorState';
+import { functionExpression_setParametersValues } from './QueryBuilderValueSpecificationModifierHelper';
+
+export const valueSpecifiation_isEqual = (
+  param1: ValueSpecification | undefined,
+  param2: ValueSpecification | undefined,
+): boolean => {
+  if (
+    param1 instanceof VariableExpression &&
+    param2 instanceof VariableExpression
+  ) {
+    return param1.name === param2.name;
+  } else if (
+    param1 instanceof PrimitiveInstanceValue &&
+    param2 instanceof PrimitiveInstanceValue
+  ) {
+    if (
+      param1.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE &&
+      param2.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE
+    ) {
+      return true;
+    } else if (
+      param1.genericType.value.rawType.name ===
+      param2.genericType.value.rawType.name
+    ) {
+      return param1.values[0] === param2.values[0];
+    }
+    return false;
+  }
+  return false;
+};
+
+export const getSourceTemporalStereotype = (
+  property: DerivedProperty,
+  graph: PureModel,
+): MILESTONING_STEROTYPE | undefined => {
+  const owner = property.owner;
+  if (owner instanceof Class) {
+    return getMilestoneTemporalStereotype(owner, graph);
+  } else if (owner instanceof Association) {
+    if (owner._generatedMilestonedProperties.length) {
+      const ownerClass =
+        owner._generatedMilestonedProperties[0]?.genericType.value.rawType;
+      return getMilestoneTemporalStereotype(
+        guaranteeType(ownerClass, Class),
+        graph,
+      );
+    }
+  }
+  return undefined;
+};
+
+//sourceStereotype: Stereotype of source class of the previous property expression/ class which we are querying from if it is the
+// first expression getting processed.
+//targetStereotype: Stereotype of source class of current property expression.
+export const isDefaultDatePropagationSupported = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  graph: PureModel,
+  prevPropertyExpression?:
+    | QueryBuilderDerivedPropertyExpressionState
+    | undefined,
+): boolean => {
+  const property = derivedPropertyExpressionState.derivedProperty;
+  if (prevPropertyExpression?.parameterValues.length) {
+    return false;
+  }
+  if (property.genericType.value.rawType instanceof Class) {
+    const sourceStereotype = getSourceTemporalStereotype(property, graph);
+    if (sourceStereotype === MILESTONING_STEROTYPE.BITEMPORAL) {
+      return true;
+    }
+    const targetStereotype = getMilestoneTemporalStereotype(
+      property.genericType.value.rawType,
+      graph,
+    );
+    return sourceStereotype === targetStereotype;
+  }
+  return false;
+};
+
+export const removePropagatedDates = (
```

Comment:
Let's add documentation to this helper method. Also, I need to check with you why we're doing this. In my mind, we can leave all of this as is if we use `propagated value` for these parameter. Now, we only need to `prune` these when we `build the lambda` for persistence/execution.